File: packages/js/src/sandbox/process.ts
import { ConnectionConfig } from '../config'
import { handleApiError } from '../errors'

/** A process this sandbox is currently running. */
export interface ProcessInfo {
  pid: number
  cmd: string[]
  /** When the process started, unix milliseconds. */
  startedAt: number
}

/** Signals accepted by {@link Process.kill}. */
export type ProcessSignal =
  | 'SIGTERM' | 'SIGKILL' | 'SIGINT' | 'SIGHUP' | 'SIGQUIT' | 'SIGUSR1' | 'SIGUSR2'

/**
 * The result of a process execution inside a Lizard microVM.
 */
export interface ProcessResult {
  stdout: string
  stderr: string
  exitCode: number
}

export interface ProcessOpts {
  timeoutMs?: number
  envs?: Record<string, string>
  user?: string
  workdir?: string
  /** Called with each stdout line as it is produced, rather than at the end. */
  onStdout?: (data: string) => void
  /** Called with each stderr line as it is produced, rather than at the end. */
  onStderr?: (data: string) => void
  /**
   * Called once with the process's pid, before any output. Gives a streaming
   * `exec` something to pass to {@link Process.kill}.
   */
  onPid?: (pid: number) => void
}

/**
 * Runs processes inside a Lizard sandbox microVM.
 *
 * Access via `sandbox.process`.
 */
export class Process {
  constructor(
    private readonly sandboxId: string,
    private readonly config: ConnectionConfig
  ) {}

  /**
   * Execute a command inside the microVM and wait for it to complete.
   *
   * The command runs in a shell inside the Lizard sandbox and returns
   * stdout, stderr, and the exit code when it finishes.
   *
   * @param cmd Shell command to run inside the microVM.
   * @param opts Optional execution options — environment variables, working
   *   directory, user, and timeout.
   *
   * @example
   * ```ts
   * const result = await sandbox.process.exec('node index.js')
   * console.log(result.stdout)
   * ```
   *
   * @example Run with a custom working directory and env vars:
   * ```ts
   * const result = await sandbox.process.exec('npm test', {
   *   workdir: '/app',
   *   envs: { NODE_ENV: 'test' },
   * })
   * ```
   */
  async exec(cmd: string, opts?: ProcessOpts): Promise<ProcessResult> {
    // `onStdout`/`onStderr` have been in ProcessOpts since the beginning but were
    // never wired up — passing them did nothing and the caller waited for the whole
    // command regardless. The platform has always streamed the output; it just needs
    // to be asked for it with an SSE Accept header.
    const streaming = Boolean(opts?.onStdout || opts?.onStderr || opts?.onPid)

    const res = await fetch(`${this.config.apiUrl}/api/sandboxes/${this.sandboxId}/exec`, {
      method: 'POST',
      headers: streaming
        ? { ...this.config.headers, Accept: 'text/event-stream' }
        : this.config.headers,
      body: JSON.stringify({
        cmd,
        envs: opts?.envs,
        user: opts?.user,
        workdir: opts?.workdir,
        timeoutMs: opts?.timeoutMs,
      }),
      signal: AbortSignal.timeout(opts?.timeoutMs ?? 60_000),
    })

    if (!res.ok) await handleApiError(res)
    if (!streaming) return res.json() as Promise<ProcessResult>
    return this.consumeStream(res, opts!)
  }

  /**
   * Read an SSE exec stream, handing each line to the caller as it arrives while
   * still accumulating the full result. Events are `{stream, line}` for output and
   * `{exitCode}` at the end.
   */
  /**
   * List the processes this sandbox is currently running.
   *
   * Only processes started through `exec` — not every process in the guest.
   * Those are the ones you can act on; the rest are the image's own business.
   *
   * @example
   * ```ts
   * for (const p of await sandbox.process.list()) {
   *   console.log(p.pid, p.cmd.join(' '))
   * }
   * ```
   */
  async list(): Promise<ProcessInfo[]> {
    const res = await fetch(`${this.config.apiUrl}/api/sandboxes/${this.sandboxId}/processes`, {
      headers: this.config.headers,
    })
    if (!res.ok) await handleApiError(res)
    return res.json() as Promise<ProcessInfo[]>
  }

  /**
   * Signal a running process. Defaults to `SIGTERM`.
   *
   * The signal goes to the process group, so a shell's children die with it —
   * killing `sh -c 'sleep 100'` otherwise leaves the sleep running.
   *
   * The pid comes from {@link list}, or from the `pid` event at the start of a
   * streaming `exec`.
   *
   * @example Stop a long build:
   * ```ts
   * const [build] = await sandbox.process.list()
   * await sandbox.process.kill(build.pid)
   * ```
   */
  async kill(pid: number, signal: ProcessSignal = 'SIGTERM'): Promise<void> {
    const res = await fetch(`${this.config.apiUrl}/api/sandboxes/${this.sandboxId}/processes/signal`, {
      method: 'POST',
      headers: this.config.headers,
      body: JSON.stringify({ pid, signal }),
    })
    if (!res.ok) await handleApiError(res)
  }

  private async consumeStream(res: Response, opts: ProcessOpts): Promise<ProcessResult> {
    const reader = res.body?.getReader()
    if (!reader) return { stdout: '', stderr: '', exitCode: 0 }

    const decoder = new TextDecoder()
    let buf = ''
    let stdout = ''
    let stderr = ''
    let exitCode = 0

    const handle = (raw: string) => {
      if (!raw.startsWith('data:')) return
      let ev: { stream?: string; line?: string; exitCode?: number; pid?: number }
      try {
        ev = JSON.parse(raw.slice(5).trim())
      } catch {
        return // a partial or non-JSON keepalive frame
      }
      if (ev.line !== undefined) {
        if (ev.stream === 'stderr') {
          stderr += ev.line + '\n'
          opts.onStderr?.(ev.line)
        } else {
          stdout += ev.line + '\n'
          opts.onStdout?.(ev.line)
        }
      }
      if (ev.pid !== undefined) opts.onPid?.(ev.pid)
      if (ev.exitCode !== undefined) exitCode = ev.exitCode
    }

    for (;;) {
      const { done, value } = await reader.read()
      if (done) break
      buf += decoder.decode(value, { stream: true })
      // Keep the trailing fragment: a chunk boundary can land mid-line, and parsing
      // half an event would drop it silently.
      const lines = buf.split('\n')
      buf = lines.pop() ?? ''
      for (const l of lines) handle(l)
    }
    if (buf) handle(buf)

    return { stdout: stdout.trimEnd(), stderr: stderr.trimEnd(), exitCode }
  }
}

File: packages/python/lizard/sandbox/process.py
from __future__ import annotations
from dataclasses import dataclass
from typing import Callable, TYPE_CHECKING

if TYPE_CHECKING:
    from ..config import ConnectionConfig, HTTP_TIMEOUT_S


@dataclass
class ProcessResult:
    """Result of a process executed inside a Lizard sandbox microVM."""

    stdout: str
    stderr: str
    exit_code: int


@dataclass
class ProcessInfo:
    """A process the sandbox is currently running."""

    pid: int
    cmd: list[str]
    #: When the process started, unix milliseconds.
    started_at: int


class Process:
    """
    Run processes inside a Lizard sandbox microVM.

    Access via ``sandbox.process``.
    """

    def __init__(self, sandbox_id: str, config: "ConnectionConfig"):
        self._sandbox_id = sandbox_id
        self._config = config

    def exec_(
        self,
        cmd: str,
        *,
        envs: dict[str, str] | None = None,
        user: str | None = None,
        workdir: str | None = None,
        timeout_ms: int | None = None,
        on_stdout: "Callable[[str], None] | None" = None,
        on_stderr: "Callable[[str], None] | None" = None,
        on_pid: "Callable[[int], None] | None" = None,
    ) -> ProcessResult:
        """
        Execute a command inside the microVM and wait for it to finish.

        The command runs in a shell inside the Lizard sandbox and returns
        stdout, stderr, and the exit code when it completes.

        :param cmd: Shell command to run inside the microVM.
        :param envs: Additional environment variables for this execution.
        :param user: Run as this Linux user (default: ``root``).
        :param workdir: Working directory inside the microVM.
        :param timeout_ms: Execution timeout in milliseconds.
        :param on_stdout: Called with each stdout line as it is produced, rather
            than at the end. Passing it switches the call to a streaming read.
        :param on_stderr: Called with each stderr line as it is produced.
        :param on_pid: Called once with the process's pid before any output, so a
            streaming call has something to pass to :meth:`kill`.

        Example::

            result = sandbox.process.exec_("node index.js")
            print(result.stdout)

        Example with options::

            result = sandbox.process.exec_(
                "npm test",
                workdir="/app",
                envs={"NODE_ENV": "test"},
            )
        """
        import httpx

        body: dict = {"cmd": cmd}
        if envs:
            body["envs"] = envs
        if user:
            body["user"] = user
        if workdir:
            body["workdir"] = workdir
        if timeout_ms:
            body["timeoutMs"] = timeout_ms

        timeout = (timeout_ms or 60_000) / 1000

        if on_stdout is not None or on_stderr is not None or on_pid is not None:
            return self._exec_streaming(body, timeout, on_stdout, on_stderr, on_pid)

        res = httpx.post(
            f"{self._config.api_url}/api/sandboxes/{self._sandbox_id}/exec",
            headers=self._config.headers,
            json=body,
            timeout=timeout,
        )

        if not res.is_success:
            from ..errors import handle_api_error
            handle_api_error(res.status_code, res.text)

        data = res.json()
        return ProcessResult(
            stdout=data.get("stdout", ""),
            stderr=data.get("stderr", ""),
            exit_code=data.get("exitCode", 0),
        )

    def _exec_streaming(
        self,
        body: dict,
        timeout: float,
        on_stdout: "Callable[[str], None] | None",
        on_stderr: "Callable[[str], None] | None",
        on_pid: "Callable[[int], None] | None" = None,
    ) -> ProcessResult:
        """Read the exec output as an SSE stream, handing each line to the caller as
        it arrives while still accumulating the full result.

        The platform has always streamed exec output; it just has to be asked for it
        with an SSE Accept header. Events are ``{"stream", "line"}`` for output and
        ``{"exitCode"}`` at the end.
        """
        import httpx
        import json as _json

        headers = dict(self._config.headers)
        headers["Accept"] = "text/event-stream"

        stdout_parts: list[str] = []
        stderr_parts: list[str] = []
        exit_code = 0

        with httpx.stream(
            "POST",
            f"{self._config.api_url}/api/sandboxes/{self._sandbox_id}/exec",
            headers=headers,
            json=body,
            timeout=timeout,
        ) as res:
            if not res.is_success:
                res.read()
                from ..errors import handle_api_error
                handle_api_error(res.status_code, res.text)
            for raw in res.iter_lines():
                if not raw.startswith("data:"):
                    continue
                try:
                    ev = _json.loads(raw[5:].strip())
                except ValueError:
                    continue  # a partial or non-JSON keepalive frame
                line = ev.get("line")
                if line is not None:
                    if ev.get("stream") == "stderr":
                        stderr_parts.append(line)
                        if on_stderr is not None:
                            on_stderr(line)
                    else:
                        stdout_parts.append(line)
                        if on_stdout is not None:
                            on_stdout(line)
                if ev.get("pid") is not None and on_pid is not None:
                    on_pid(ev["pid"])
                if ev.get("exitCode") is not None:
                    exit_code = ev["exitCode"]

        return ProcessResult(
            stdout="\n".join(stdout_parts),
            stderr="\n".join(stderr_parts),
            exit_code=exit_code,
        )

    def list(self) -> list[ProcessInfo]:
        """List the processes this sandbox is currently running.

        Only processes started through :meth:`exec_` -- not every process in the
        guest. Those are the ones you can act on; the rest are the image's own
        business.
        """
        import httpx

        res = httpx.get(
            f"{self._config.api_url}/api/sandboxes/{self._sandbox_id}/processes",
            headers=self._config.headers,
            timeout=HTTP_TIMEOUT_S,
        )
        if not res.is_success:
            from ..errors import handle_api_error
            handle_api_error(res.status_code, res.text)
        return [
            ProcessInfo(pid=p["pid"], cmd=p.get("cmd", []), started_at=p.get("startedAt", 0))
            for p in res.json()
        ]

    def kill(self, pid: int, signal: str = "SIGTERM") -> None:
        """Signal a running process. Defaults to ``SIGTERM``.

        The signal goes to the process group, so a shell's children die with it --
        killing ``sh -c 'sleep 100'`` otherwise leaves the sleep running.

        The pid comes from :meth:`list`, or from ``on_pid`` during a streaming
        :meth:`exec_`.
        """
        import httpx

        res = httpx.post(
            f"{self._config.api_url}/api/sandboxes/{self._sandbox_id}/processes/signal",
            headers=self._config.headers,
            json={"pid": pid, "signal": signal},
            timeout=HTTP_TIMEOUT_S,
        )
        if not res.is_success:
            from ..errors import handle_api_error
            handle_api_error(res.status_code, res.text)
